crates/compute/src/cpu/layer.rs

```diff
@@ -50,4 +62,276 @@ impl<F: TowerField> ComputeLayer<F> for CpuLayer<F> {
 		dst.copy_from_slice(src);
 		Ok(())
 	}
+
+	fn execute(
+		&self,
+		f: impl FnOnce(&mut Self::Exec) -> Result<Vec<T::B128>, Error>,
+	) -> Result<Vec<T::B128>, Error> {
+		f(&mut CpuExecutor)
+	}
+
+	fn inner_product<'a>(
+		&'a self,
+		_exec: &'a mut Self::Exec,
+		a_edeg: usize,
+		a_in: &'a [T::B128],
+		b_in: &'a [T::B128],
+	) -> Result<T::B128, Error> {
+		if a_edeg > T::B128::TOWER_LEVEL
+			|| a_in.len() << (T::B128::TOWER_LEVEL - a_edeg) != b_in.len()
+		{
+			return Err(Error::InputValidation(format!(
+				"invalid input: a_edeg={a_edeg} |a|={} |b|={}",
+				a_in.len(),
+				b_in.len()
+			)));
+		}
+
+		let result = match a_edeg {
+			0 => inner_product_unchecked(
+				b_in.iter().copied(),
+				a_in.iter()
+					.flat_map(<T::B128 as ExtensionField<T::B1>>::iter_bases),
+			),
+			3 => inner_product_unchecked(
+				b_in.iter().copied(),
+				a_in.iter()
+					.flat_map(<T::B128 as ExtensionField<T::B8>>::iter_bases),
+			),
+			4 => inner_product_unchecked(
+				b_in.iter().copied(),
+				a_in.iter()
+					.flat_map(<T::B128 as ExtensionField<T::B16>>::iter_bases),
+			),
+			5 => inner_product_unchecked(
+				b_in.iter().copied(),
+				a_in.iter()
+					.flat_map(<T::B128 as ExtensionField<T::B32>>::iter_bases),
+			),
+			6 => inner_product_unchecked(
+				b_in.iter().copied(),
+				a_in.iter()
+					.flat_map(<T::B128 as ExtensionField<T::B64>>::iter_bases),
+			),
+			7 => inner_product_unchecked::<T::B128, T::B128>(
+				a_in.iter().copied(),
+				b_in.iter().copied(),
+			),
+			_ => {
+				return Err(Error::InputValidation(format!(
+					"unsupported value of a_edeg: {a_edeg}"
+				)))
+			}
+		};
+		Ok(result)
+	}
+
+	fn tensor_expand(
+		&self,
+		_exec: &mut Self::Exec,
+		log_n: usize,
+		coordinates: &[T::B128],
+		data: &mut &mut [T::B128],
+	) -> Result<(), Error> {
+		if data.len() != 1 << (log_n + coordinates.len()) {
+			return Err(Error::InputValidation(format!("invalid data length: {}", data.len())));
+		}
+
+		for (i, r_i) in coordinates.iter().enumerate() {
+			let (lhs, rest) = data.split_at_mut(1 << (log_n + i));
+			let (rhs, _rest) = rest.split_at_mut(1 << (log_n + i));
+			for (x_i, y_i) in std::iter::zip(lhs, rhs) {
+				let prod = *x_i * r_i;
+				*x_i -= prod;
+				*y_i += prod;
+			}
+		}
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use std::iter::repeat_with;
+
+	use binius_field::{
+		BinaryField128b, BinaryField16b, BinaryField32b, ExtensionField, Field, PackedExtension,
+		PackedField, TowerField,
+	};
+	use binius_math::{tensor_prod_eq_ind, MultilinearExtension, MultilinearQuery};
+	use bytemuck::zeroed_vec;
+	use rand::{prelude::StdRng, SeedableRng};
+
+	use super::*;
+	use crate::{memory::ComputeMemory, tower::CanonicalTowerFamily};
+
+	#[test]
+	fn test_exec_single_tensor_expand() {
+		let n_vars = 8;
+		let mut rng = StdRng::seed_from_u64(0);
+		let coordinates = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
```

Comment:
More of this code (ideally almost everything except creating the `hal`) should go into the shared test. Use `copy_h2d` to copy to a device buffer. Same with the other tests. See the memory tests in `src/layer.rs` for example.